numpy/random/mtrand/mtrand.pyx

```diff
@@ -384,7 +389,7 @@ cdef object discnp_array(rk_state *state, rk_discnp func, object size,
     else:
         array = <ndarray>np.empty(size, dtype=int)
         multi = <broadcast>np.broadcast(on, op, array)
-        if multi.shape != array.shape:
+        if multi.shape != getattr(array, 'shape'):
```

Comment:
Do `x.attr` and `getattr(x, 'attr')` behave differently in cython? If so, that's pretty scary.

Reply:
`array.shape` becomes [`array.dimensions`](https://github.com/numpy/numpy/pull/12284/files#diff-35cf7adc200725d2d15266c5e58a07f3R253) in C, but `multi.shape` is `PyObject_GetAttrString(multi, "shape")`, so we either need to make sure cython uses `PyObject_GetAttrString(array, "shape")` and not the fast C struct field reference.

Reply:
If we ask for the non-deprecated API in this file, doesn't that resolve the problem by making Cython unaware of the underlying struct members?

Reply:
Using the non-deprecated API would require converting `array.shape` to `npy_intp* PyArray_SHAPE(array)`. cython/cython#2640 aims to automatically do the conversion. This would still not return a `PyObject*`.

Reply:
That is an existing issue/feature of cython, not part of any changes made for this PR. For instance, it is worked around in scipy [here](https://github.com/scipy/scipy/blob/v1.1.0/scipy/ndimage/src/_ni_label.pyx#L210) which shows a more cythonic construct than the one I used but produces the same C code. I will change this PR to use `(<object>array).shape` casting.